### ISSUE TITLE
Allows binding loopback address only

### ISSUE THREAD

**Issue by gzz2000** [beringresearch/macpine#105]
Currently, macpine always binds some public SSH port on all interfaces, and I could not find a way to attach to local interface only. I currently have to disable qemu's incoming connection in my firewall. 

It is theoretically not a difficult thing to do. All we have to do is to specify a complete host IP in qemu settings: https://github.com/beringresearch/macpine/search?q=hostfwd

By the way I love this project and it helps me a lot.

**Comment by maxzinkus** [comment]
Theoretically this is not difficult, you are correct. Unfortunately in this case, theory and practice are only the same in theory. In practice, adding `tcp:127.0.0.1: ...` does not make `qemu` correctly bind localhost and limit accessibility to loopback.

I've [opened an upstream issue](https://gitlab.com/qemu-project/qemu/-/issues/1560) on `qemu` if you want to go upvote it and comment with your support.

**Comment by maxzinkus** [comment]
For tracking: Blocked by https://gitlab.com/qemu-project/qemu/-/issues/1560

**Comment by maxzinkus** [comment]
@idroz you could add [Blocked] to the title of this issue and we can sit on it until the upstream is resolved/closed :)

**Comment by maxzinkus** [comment]
Update! Upstream issue is closed, and it's Way Weirder than I expected. It's not actually a QEMU bug -- macOS doesn't let you bind `127.0.0.1:x` for ports `x < 1024` without root privileges. This is standard unix behavior.

BUT

macOS _will_ let you bind `*:x` for any privileged port `x` without root.

Tested it with `netcat`:

```bash
$ nc -l 22
... no output, works, listening on 22 for any connections ...
```

whereas

```bash
$ nc -l 127.0.0.1 22
nc: permission denied
```

and

```bash
$ sudo nc -l 127.0.0.1 22
... works ...
```

So, defaulting to 22 for macpine ssh has been working because `qemu-system-` can bind `*:22` and redirect it to the VM. But if you put in `tcp:127.0.0.1:22` qemu (really, [slirp](https://gitlab.freedesktop.org/slirp/libslirp) under the hood) tries to bind that address -- and fails due to lacking root privileges.

I see the value of binding only on localhost for security.

Therefore, there are two approaches to add this use case, in addition to keeping the current behavior as default:

- run macpine as root -> unacceptable for security, convenience, so many reasons
- if a user specifies local-only forwarding, default to SSH ports above 1024, and basically tell users in docs that if they want privileged ports forwarded in, they should set up a privileged listener on the host that redirects to the unprivileged qemu-forwarded port. Something like socat:

`sudo socat TCP4-LISTEN:<privileged port> TCP4:localhost:<qemu-forwarded port>`

e.g.

`sudo socat TCP4-LISTEN:22 TCP4:localhost:20022` with the macpine VM `SSHPort` set to 20022.

**Comment by maxzinkus** [comment]
So, I propose an optional configuration in launch and in `config.yaml` (perhaps `-l / --local-fwd`) which forces port bindings (`SSHPort` and other forwarded `Ports`) to bind `127.0.0.1`, becoming only available over the loopback interface.

- If they are privileged ports and qemu can't bind them, we simply fail out and propagate up the permission denied error.
- If macpine is for some reason being run as root, then just let whatever happens happen.

@idroz what do you think of this?

**Comment by gzz2000** [comment]
Wow I didn't expect such a reason. For the security concerns, it might be interesting to look at this:
https://github.com/Castaglia/MacOSX-authbind

It is a more transparent solution than socat. Although it can be hacky to bundle that with macpine.

**Comment by maxzinkus** [comment]
Yeah, I've seen authbind. I think your intuition is right that it'll be pretty hacky to depend on an old, smaller project like that, and to add a system capability to a binary that is already a security point-of-concern (VM escapes can utilize that privilege if they exploit qemu).

I'm hesitant to work around system protections of privileged ports.

As it stands, qemu triggers a prompt to allow incoming connections from the local network if the macOS firewall is configured. If security is a concern, the firewall should be enabled, and if this permission is denied, local loopback connections can still be made to the `*:22` listener.

Therefore, as there are system-level protections against misuse of the `*:22` port binding and qemu is currently able to run unprivileged in a seamless way, I think the best response will be to close this as `wontfix` despite the annoyance `*` binds creates.

Thanks for your report, it was definitely an issue in need of discussion.

I'll add to the documentation to summarize these points.

**Comment by maxzinkus** [comment]
@idroz if you have any opinions on the above, let me know! :)
